packages/auth-remix/src/server.ts

```diff
@@ -0,0 +1,795 @@
+                }),
+              }
+            );
+          }
+          case "oauth/callback": {
+            if (!onOAuthCallback) {
+              throw new Error(
+                `'onOAuthCallback' auth route handler not configured`
+              );
+            }
+            const error = searchParams.get("error");
+            if (error) {
+              const desc = searchParams.get("error_description");
+              return onOAuthCallback({
+                error: new Error(error + (desc ? `: ${desc}` : "")),
+              });
+            }
+            const code = searchParams.get("code");
+            const isSignUp = searchParams.get("isSignUp") === "true";
+
+            const verifier =
+              parseCookies(request)[this.options.pkceVerifierCookieName];
+
+            if (!code) {
+              return onOAuthCallback({
+                error: new Error("no pkce code in response"),
+              });
+            }
+
+            if (!verifier) {
+              return onOAuthCallback({
+                error: new Error("no pkce verifier cookie found"),
+              });
+            }
+            let tokenData: TokenData;
+            try {
+              tokenData = await (await this.core).getToken(code, verifier);
+            } catch (err) {
+              return onOAuthCallback({
+                error: err instanceof Error ? err : new Error(String(err)),
+              });
+            }
+
+            const headers = new Headers();
+            headers.append(
+              "Set-Cookie",
+              `${this.options.authCookieName}=${tokenData.auth_token}; HttpOnly; SameSite=Lax; Path=/`
+            );
+            headers.append(
+              "Set-Cookie",
+              `${this.options.pkceVerifierCookieName}=`
+            );
+
+            return oathCallbackCall(onOAuthCallback, headers, {
+              error: null,
+              tokenData,
+              provider: searchParams.get(
+                "provider"
+              ) as BuiltinOAuthProviderNames,
+              isSignUp,
+            });
+          }
+          case "builtin/callback": {
+            if (!onBuiltinUICallback) {
+              throw new Error(
+                `'onBuiltinUICallback' auth route handler not configured`
+              );
+            }
+            const error = searchParams.get("error");
+            if (error) {
+              const desc = searchParams.get("error_description");
+              return onBuiltinUICallback({
+                error: new Error(error + (desc ? `: ${desc}` : "")),
+              });
+            }
+            const code = searchParams.get("code");
+            const verificationEmailSentAt = searchParams.get(
+              "verification_email_sent_at"
+            );
+
+            if (!code) {
+              if (verificationEmailSentAt) {
+                return onBuiltinUICallback({
```

Comment:
If we're checking that the responses returned to callbacks are always redirects, we need to wrap all these callback calls in that check.